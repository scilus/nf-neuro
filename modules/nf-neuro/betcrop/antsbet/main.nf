
process BETCROP_ANTSBET {
    tag "$meta.id"
    label 'process_high'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_2.0.2.sif':
        'scilus/scilus:2.0.2' }"

    input:
    tuple val(meta), path(t1), path(template), path(tissues_probabilities), path(mask), path(initial_affine)

    output:
    tuple val(meta), path("*t1_bet.nii.gz")     , emit: t1
    tuple val(meta), path("*t1_bet_mask.nii.gz"), emit: mask
    path "versions.yml"                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def args = []
    if (mask) args += ["-f $mask"]
    if (initial_affine) args += ["-r $initial_affine"]

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=$task.cpus
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1
    export ANTS_RANDOM_SEED=1234

    antsBrainExtraction.sh -d 3 -a $t1 -o bet/ -u 0 \
        -e $template -m $tissues_probabilities ${args.join(' ')}
    scil_volume_math.py convert bet/BrainExtractionMask.nii.gz \
        ${prefix}__t1_bet_mask.nii.gz --data_type uint8
    scil_volume_math.py multiplication $t1 ${prefix}__t1_bet_mask.nii.gz \
        ${prefix}__t1_bet.nii.gz --data_type float32

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
        ants: \$(antsRegistration --version | grep "Version" | sed -E 's/.*v([0-9]+\\+\\).*/\\1/')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """

    touch ${prefix}__t1_bet.nii.gz
    touch ${prefix}__t1_bet_mask.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
        ants: \$(antsRegistration --version | grep "Version" | sed -E 's/.*v([0-9]+\\+\\).*/\\1/')
    END_VERSIONS

    function handle_code () {
    local code=\$?
    ignore=( 1 )
    exit \$([[ " \${ignore[@]} " =~ " \$code " ]] && echo 0 || echo \$code)
    }
    trap 'handle_code' ERR

    antsBrainExtraction.sh
    scil_volume_math.py -h

    """
}
