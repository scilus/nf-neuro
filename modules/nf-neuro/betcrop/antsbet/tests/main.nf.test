nextflow_process {

    name "Test Process BETCROP_ANTSBET"
    script "../main.nf"
    process "BETCROP_ANTSBET"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "betcrop"
    tag "betcrop/antsbet"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    test("betcrop - antsbet") {
        setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "T1w.zip", "transform.zip", "antsbet.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1: it.simpleName == "T1w"
                        transform: it.simpleName == "transform"
                        template: it.simpleName == "antsbet"
                    }
                ch_t1 = ch_split_test_data.t1.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]
                }
                ch_template = ch_split_test_data.template.map{
                    test_data_directory -> [
                        [ id: 'test' ],
                        file("\${test_data_directory}/t1_template.nii.gz"),
                        file("\${test_data_directory}/t1_brain_probability_map.nii.gz")
                    ]
                }
                ch_mask = ch_split_test_data.transform.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_to_bet_template_mask.nii.gz")
                    ]
                }
                ch_transform = ch_split_test_data.transform.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_to_bet_template.mat")
                    ]
                }
                input[0] = ch_t1
                    .join(ch_template)
                    .join(ch_mask)
                    .join(ch_transform)
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    niftiMD5SUM(process.out.t1.get(0).get(1), 1),
                    niftiMD5SUM(process.out.mask.get(0).get(1)),
                    process.out.versions
                ).match() }
            )
        }
    }
    test("betcrop - antsbet - stub-run") {
        options "-stub-run"
        setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "T1w.zip", "transform.zip", "antsbet.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1: it.simpleName == "T1w"
                        transform: it.simpleName == "transform"
                        template: it.simpleName == "antsbet"
                    }
                ch_t1 = ch_split_test_data.t1.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]
                }
                ch_template = ch_split_test_data.template.map{
                    test_data_directory -> [
                        [ id: 'test' ],
                        file("\${test_data_directory}/t1_template.nii.gz"),
                        file("\${test_data_directory}/t1_brain_probability_map.nii.gz")
                    ]
                }
                ch_mask = ch_split_test_data.transform.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_to_bet_template_mask.nii.gz")
                    ]
                }
                ch_transform = ch_split_test_data.transform.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_to_bet_template.mat")
                    ]
                }
                input[0] = ch_t1
                    .join(ch_template)
                    .join(ch_mask)
                    .join(ch_transform)
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match() }
            )
        }
    }
}
