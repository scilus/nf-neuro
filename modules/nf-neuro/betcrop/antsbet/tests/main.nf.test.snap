{
    "betcrop - antsbet": {
        "content": [
            "test__t1_bet.nii.gz:md5:header,e7cfbd06624321d70cbd667a77315ba3,data,381da559c1f38a2526f58b741eb0f7fc",
            "test__t1_bet_mask.nii.gz:md5:header,f7389fe98c9a7e3a87c90b7ca05ea14a,data,1572808125554e50ff73fbe0e28037a9",
            [
                "versions.yml:md5,bb378e913d4002d0b644d3277892ae20"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-01-15T16:55:54.33351491"
    },
    "betcrop - antsbet - stub-run": {
        "content": [
            [
                "versions.yml:md5,bb378e913d4002d0b644d3277892ae20"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-01-14T20:42:47.191136661"
    }
}